pkg/reconciler/clusterstore/clusterstore.go

```diff
@@ -89,37 +97,57 @@ func (c *Reconciler) updateClusterStoreStatus(ctx context.Context, desired *v1al
 	return err
 }
 
-func (c *Reconciler) reconcileClusterStoreStatus(clusterStore *v1alpha1.ClusterStore) (*v1alpha1.ClusterStore, error) {
-	buildpacks, err := c.StoreReader.Read(clusterStore.Spec.Sources)
+func (c *Reconciler) reconcileClusterStoreStatus(ctx context.Context, clusterStore *v1alpha1.ClusterStore) (*v1alpha1.ClusterStore, error) {
+	var keychain *authn.Keychain
+
+	if clusterStore.Spec.ServiceAccountRef != nil {
+		k, err := c.KeychainFactory.KeychainForSecretRef(ctx, registry.SecretRef{
+			ServiceAccount: clusterStore.Spec.ServiceAccountRef.Name,
+			Namespace:      clusterStore.Spec.ServiceAccountRef.Namespace,
+		})
+		keychain = &k
+		if err != nil {
+			clusterStore.Status = v1alpha1.ClusterStoreStatus{
+				Status: c.createStatus(clusterStore.Generation, err),
+			}
+			return clusterStore, err
+		}
+
+	}
+	buildpacks, err := c.StoreReader.Read(keychain, clusterStore.Spec.Sources)
 	if err != nil {
 		clusterStore.Status = v1alpha1.ClusterStoreStatus{
-			Status: corev1alpha1.Status{
-				ObservedGeneration: clusterStore.Generation,
-				Conditions: corev1alpha1.Conditions{
-					{
-						Type:               corev1alpha1.ConditionReady,
-						Status:             corev1.ConditionFalse,
-						LastTransitionTime: corev1alpha1.VolatileTime{Inner: metav1.Now()},
-						Message:            err.Error(),
-					},
-				},
-			},
+			Status: c.createStatus(clusterStore.Generation, err),
 		}
 		return clusterStore, err
 	}
 
 	clusterStore.Status = v1alpha1.ClusterStoreStatus{
 		Buildpacks: buildpacks,
-		Status: corev1alpha1.Status{
-			ObservedGeneration: clusterStore.Generation,
-			Conditions: corev1alpha1.Conditions{
-				{
-					LastTransitionTime: corev1alpha1.VolatileTime{Inner: metav1.Now()},
-					Type:               corev1alpha1.ConditionReady,
-					Status:             corev1.ConditionTrue,
-				},
+		Status:     c.createStatus(clusterStore.Generation, nil),
+	}
+	return clusterStore, nil
+}
+
+func (c *Reconciler) createStatus(generation int64, err error) corev1alpha1.Status {
```

Comment:
This code is actually exactly the same for clusterstore and clusterstack. Is there a place to put such common "utility" functions?

Reply:
Utility functions that operate on the core types can probably go in the `/pkg/apis/core/v1alpha1` package.